src/coordination/coordinator_data.cpp

```diff
@@ -0,0 +1,215 @@
+
+#include <range/v3/view.hpp>
+#include <shared_mutex>
+
+namespace memgraph::coordination {
+
+CoordinatorData::CoordinatorData() {
+  auto find_instance = [](CoordinatorData *coord_data, std::string_view instance_name) -> CoordinatorInstance & {
+    std::shared_lock<utils::RWLock> lock{coord_data->coord_data_lock_};
+
+    auto instance = std::ranges::find_if(
+        coord_data->registered_instances_,
+        [instance_name](const CoordinatorInstance &instance) { return instance.InstanceName() == instance_name; });
+
+    MG_ASSERT(instance != coord_data->registered_instances_.end(), "Instance {} not found during callback!",
+              instance_name);
+    return *instance;
+  };
+
+  replica_succ_cb_ = [find_instance](CoordinatorData *coord_data, std::string_view instance_name) -> void {
+    spdlog::trace("Instance {} performing replica successful callback", instance_name);
+    auto &instance = find_instance(coord_data, instance_name);
+    MG_ASSERT(instance.IsReplica(), "Instance {} is not a replica!", instance_name);
+    instance.UpdateLastResponseTime();
+  };
+
+  replica_fail_cb_ = [find_instance](CoordinatorData *coord_data, std::string_view instance_name) -> void {
+    spdlog::trace("Instance {} performing replica failure callback", instance_name);
+    auto &instance = find_instance(coord_data, instance_name);
+    MG_ASSERT(instance.IsReplica(), "Instance {} is not a replica!", instance_name);
+    instance.UpdateInstanceStatus();
+  };
+
+  main_succ_cb_ = [find_instance](CoordinatorData *coord_data, std::string_view instance_name) -> void {
+    spdlog::trace("Instance {} performing main successful callback", instance_name);
+    auto &instance = find_instance(coord_data, instance_name);
+    MG_ASSERT(instance.IsMain(), "Instance {} is not a main!", instance_name);
+    instance.UpdateLastResponseTime();
+  };
+
+  main_fail_cb_ = [this, find_instance](CoordinatorData *coord_data, std::string_view instance_name) -> void {
+    spdlog::trace("Instance {} performing main failure callback", instance_name);
+    auto &instance = find_instance(coord_data, instance_name);
+    MG_ASSERT(instance.IsMain(), "Instance {} is not a main!", instance_name);
+    if (bool main_alive = instance.UpdateInstanceStatus(); !main_alive) {
+      spdlog::info("Main instance {} is not alive, starting automatic failover", instance_name);
+      switch (auto failover_status = DoFailover(); failover_status) {
+        using enum DoFailoverStatus;
+        case ALL_REPLICAS_DOWN:
+          spdlog::warn("Failover aborted since all replicas are down!");
+          break;
+        case MAIN_ALIVE:
+          spdlog::warn("Failover aborted since main is alive!");
+          break;
+        case RPC_FAILED:
+          spdlog::warn("Failover aborted since promoting replica to main failed!");
+          break;
+        case SUCCESS:
+          break;
+      }
+    }
+  };
+}
+
+auto CoordinatorData::DoFailover() -> DoFailoverStatus {
```

Comment:
A general comment for Failover.
REPLICAs currently ingest any data coming in from "MAIN". I don't think failover will work until we make the REPLICA-MAIN logic more robust.
The coordinator needs to label each instance and tell REPLICAs who the true MAIN is.
There could be a new field in the RPC that IDs the originating instance and REPLICAs ignore anything coming in from a different ID.

If we do failover with the current setup, we can have a situation where MAIN loses the connection to COORDINATOR, but does no lose the connection to REPLICAs and will continue sending messages to them, which they will ingest no questions asked.

Reply:
Yes 100% agree, will be handled, probably not immediately but yes

Reply:
Yes, also agree, we talked about this but want to just get feedback on generally how all this works